pkg/ccl/testccl/sqlccl/sqlccl_test: TestExplainRedactDDL failed

pkg/ccl/testccl/sqlccl/sqlccl_test.TestExplainRedactDDL [failed](https://tanzanite.cluster.engflow.com/invocations/default/997d5750-5661-4a01-b5a8-a392d28f4fcd?testReportRun=17&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvc3FsY2NsOnNxbGNjbF90ZXN0) on master @ [e0068814dfcb4f975a53b79b5546b5fb85c0f927](https://github.com/cockroachdb/cockroach/commits/e0068814dfcb4f975a53b79b5546b5fb85c0f927):

Fatal error:

```
panic: during (fs.Env).Close there remain open files; all files must be closed before closing the Env

1 open files:
github.com/cockroachdb/pebble/vfs/vfstest.(*openFilesFS).Create
 github.com/cockroachdb/pebble/vfs/vfstest/external/com_github_cockroachdb_pebble/vfs/vfstest/open_files.go:51
github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).Create.func1
 github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:741
github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).timeFilesystemOp
 github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:648
github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).Create
 github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:740
github.com/cockroachdb/pebble/vfs.(*enospcFS).Create
 github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_full.go:193
github.com/cockroachdb/cockroach/pkg/storage/fs.(*Env).Create
 github.com/cockroachdb/cockroach/pkg/storage/fs/fs.go:342
github.com/cockroachdb/cockroach/pkg/storage/fs.CreateWithSync
 github.com/cockroachdb/cockroach/pkg/storage/fs/fs.go:490
github.com/cockroachdb/cockroach/pkg/sql/colcontainer.(*diskQueue).rotateFile
 github.com/cockroachdb/cockroach/pkg/sql/colcontainer/diskqueue.go:497
github.com/cockroachdb/cockroach/pkg/sql/colcontainer.(*diskQueue).Enqueue
 github.com/cockroachdb/cockroach/pkg/sql/colcontainer/diskqueue.go:631
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils.(*SpillingQueue).Enqueue
 github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils/spilling_queue.go:215
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoiner).consumeRightInput
 github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:185
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoiner).Next
 github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:99
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Next
 github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:93
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).next
 github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:118
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
 github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Next
 github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:126
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoiner).consumeRightInput
 github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:183
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoiner).Next
 github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:99
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Next
 github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:93
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).next
 github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:118

 [recovered]
	panic: during (fs.Env).Close there remain open files; all files must be closed before closing the Env

1 open files:
github.com/cockroachdb/pebble/vfs/vfstest.(*openFilesFS).Create
 github.com/cockroachdb/pebble/vfs/vfstest/external/com_github_cockroachdb_pebble/vfs/vfstest/open_files.go:51
github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).Create.func1
 github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:741
github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).timeFilesystemOp
 github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:648
github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).Create
 github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:740
github.com/cockroachdb/pebble/vfs.(*enospcFS).Create
 github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_full.go:193
github.com/cockroachdb/cockroach/pkg/storage/fs.(*Env).Create
 github.com/cockroachdb/cockroach/pkg/storage/fs/fs.go:342
github.com/cockroachdb/cockroach/pkg/storage/fs.CreateWithSync
 github.com/cockroachdb/cockroach/pkg/storage/fs/fs.go:490
github.com/cockroachdb/cockroach/pkg/sql/colcontainer.(*diskQueue).rotateFile
 github.com/cockroachdb/cockroach/pkg/sql/colcontainer/diskqueue.go:497
github.com/cockroachdb/cockroach/pkg/sql/colcontainer.(*diskQueue).Enqueue
 github.com/cockroachdb/cockroach/pkg/sql/colcontainer/diskqueue.go:631
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils.(*SpillingQueue).Enqueue
 github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils/spilling_queue.go:215
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoiner).consumeRightInput
 github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:185
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoiner).Next
 github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:99
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Next
 github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:93
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).next
 github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:118
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
 github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Next
 github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:126
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoiner).consumeRightInput
 github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:183
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoiner).Next
 github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:99
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Next
 github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:93
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).next
 github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:118

 [recovered]
	panic: during (fs.Env).Close there remain open files; all files must be closed before closing the Env

1 open files:
github.com/cockroachdb/pebble/vfs/vfstest.(*openFilesFS).Create
 github.com/cockroachdb/pebble/vfs/vfstest/external/com_github_cockroachdb_pebble/vfs/vfstest/open_files.go:51
github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).Create.func1
 github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:741
github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).timeFilesystemOp
 github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:648
github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).Create
 github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:740
github.com/cockroachdb/pebble/vfs.(*enospcFS).Create
 github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_full.go:193
github.com/cockroachdb/cockroach/pkg/storage/fs.(*Env).Create
 github.com/cockroachdb/cockroach/pkg/storage/fs/fs.go:342
github.com/cockroachdb/cockroach/pkg/storage/fs.CreateWithSync
 github.com/cockroachdb/cockroach/pkg/storage/fs/fs.go:490
github.com/cockroachdb/cockroach/pkg/sql/colcontainer.(*diskQueue).rotateFile
 github.com/cockroachdb/cockroach/pkg/sql/colcontainer/diskqueue.go:497
github.com/cockroachdb/cockroach/pkg/sql/colcontainer.(*diskQueue).Enqueue
 github.com/cockroachdb/cockroach/pkg/sql/colcontainer/diskqueue.go:631
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils.(*SpillingQueue).Enqueue
 github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils/spilling_queue.go:215
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoiner).consumeRightInput
 github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:185
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoiner).Next
 github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:99
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Next
 github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:93
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).next
 github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:118
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
 github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Next
 github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:126
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoiner).consumeRightInput
 github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:183
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoiner).Next
 github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:99
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Next
 github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:93
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).next
 github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:118

 [recovered]
	panic: during (fs.Env).Close there remain open files; all files must be closed before closing the Env

1 open files:
github.com/cockroachdb/pebble/vfs/vfstest.(*openFilesFS).Create
 github.com/cockroachdb/pebble/vfs/vfstest/external/com_github_cockroachdb_pebble/vfs/vfstest/open_files.go:51
github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).Create.func1
 github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:741
github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).timeFilesystemOp
 github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:648
github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).Create
 github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:740
github.com/cockroachdb/pebble/vfs.(*enospcFS).Create
 github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_full.go:193
github.com/cockroachdb/cockroach/pkg/storage/fs.(*Env).Create
 github.com/cockroachdb/cockroach/pkg/storage/fs/fs.go:342
github.com/cockroachdb/cockroach/pkg/storage/fs.CreateWithSync
 github.com/cockroachdb/cockroach/pkg/storage/fs/fs.go:490
github.com/cockroachdb/cockroach/pkg/sql/colcontainer.(*diskQueue).rotateFile
 github.com/cockroachdb/cockroach/pkg/sql/colcontainer/diskqueue.go:497
github.com/cockroachdb/cockroach/pkg/sql/colcontainer.(*diskQueue).Enqueue
 github.com/cockroachdb/cockroach/pkg/sql/colcontainer/diskqueue.go:631
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils.(*SpillingQueue).Enqueue
 github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils/spilling_queue.go:215
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoiner).consumeRightInput
 github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:185
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoiner).Next
 github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:99
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Next
 github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:93
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).next
 github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:118
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
 github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Next
 github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:126
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoiner).consumeRightInput
 github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:183
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoiner).Next
 github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:99
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Next
 github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:93
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).next
 github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:118



```
Stack: 

```
goroutine 16 [running]:
testing.tRunner.func1.2({0x5ff99e0, 0xc008eb0bf0})
	GOROOT/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1634 +0x377
panic({0x5ff99e0?, 0xc008eb0bf0?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:133 +0x38a
panic({0x5ff99e0?, 0xc008eb0bf0?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x111edea?, {0x7ee2568, 0xc2da240})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x65
panic({0x5ff99e0?, 0xc008eb0bf0?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/storage/fs.InitEnv.func1()
	github.com/cockroachdb/cockroach/pkg/storage/fs/fs.go:158 +0xd8
github.com/cockroachdb/cockroach/pkg/storage/fs.(*Env).Close(0xc0054a62d0)
	github.com/cockroachdb/cockroach/pkg/storage/fs/fs.go:299 +0x2ac
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleTempEngine).Close(0xc001a6d9f0)
	github.com/cockroachdb/cockroach/pkg/storage/temp_engine.go:45 +0x96
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc003d6d4d0, {0x7ee2568, 0xc2da240})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:550 +0x131
pkg/ccl/testccl/sqlccl/sqlccl_test.TestExplainRedactDDL(0xc005114680)
	pkg/ccl/testccl/sqlccl/sqlccl_test/pkg/ccl/testccl/sqlccl/explain_test.go:129 +0xa3e
testing.tRunner(0xc005114680, 0x6a86bd8)
	GOROOT/src/testing/testing.go:1689 +0xfb
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1742 +0x390
```
<details><summary>Log preceding fatal error</summary>
<p>

```
*   | testing.tRunner
*   | 	GOROOT/src/testing/testing.go:1689
*   | runtime.goexit
*   | 	src/runtime/asm_amd64.s:1695
* Wraps: (4) during (fs.Env).Close there remain open files; all files must be closed before closing the Env
*   |
*   | 1 open files:
*   | github.com/cockroachdb/pebble/vfs/vfstest.(*openFilesFS).Create
*   |  github.com/cockroachdb/pebble/vfs/vfstest/external/com_github_cockroachdb_pebble/vfs/vfstest/open_files.go:51
*   | github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).Create.func1
*   |  github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:741
*   | github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).timeFilesystemOp
*   |  github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:648
*   | github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).Create
*   |  github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:740
*   | github.com/cockroachdb/pebble/vfs.(*enospcFS).Create
*   |  github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_full.go:193
*   | github.com/cockroachdb/cockroach/pkg/storage/fs.(*Env).Create
*   |  github.com/cockroachdb/cockroach/pkg/storage/fs/fs.go:342
*   | github.com/cockroachdb/cockroach/pkg/storage/fs.CreateWithSync
*   |  github.com/cockroachdb/cockroach/pkg/storage/fs/fs.go:490
*   | github.com/cockroachdb/cockroach/pkg/sql/colcontainer.(*diskQueue).rotateFile
*   |  github.com/cockroachdb/cockroach/pkg/sql/colcontainer/diskqueue.go:497
*   | github.com/cockroachdb/cockroach/pkg/sql/colcontainer.(*diskQueue).Enqueue
*   |  github.com/cockroachdb/cockroach/pkg/sql/colcontainer/diskqueue.go:631
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils.(*SpillingQueue).Enqueue
*   |  github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils/spilling_queue.go:215
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoiner).consumeRightInput
*   |  github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:185
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoiner).Next
*   |  github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:99
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Next
*   |  github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:93
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).next
*   |  github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:118
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
*   |  github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Next
*   |  github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:126
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoiner).consumeRightInput
*   |  github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:183
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoiner).Next
*   |  github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.go:99
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Next
*   |  github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:93
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).next
*   |  github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:118
* Error types: (1) *withstack.withStack (2) *assert.withAssertionFailure (3) *withstack.withStack (4) *errutil.leafError
*
            --- PASS: TestExplainRedactDDL/ExplainAnalyze/Ddl/Viz (0.01s)
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=17</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #121545 pkg/ccl/testccl/sqlccl/sqlccl_test: TestExplainRedactDDL failed [C-test-failure O-robot P-2 T-sql-queries branch-release-23.1.18-rc]
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExplainRedactDDL.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-38054

--- a/pkg/sql/colcontainer/diskqueue.go
+++ b/pkg/sql/colcontainer/diskqueue.go
@@ -440,8 +440,15 @@ func (d *diskQueue) closeFileDeserializer(ctx context.Context) error {
 	return nil
 }
 
-func (d *diskQueue) Close(ctx context.Context) error {
+func (d *diskQueue) Close(ctx context.Context) (retErr error) {
 	defer func() {
+		if d.writeFile != nil {
+			// Ensure that we always attempt to close the file in case we
+			// short-circuit the method due to an error.
+			if err := d.writeFile.Close(); err != nil {
+				retErr = errors.CombineErrors(retErr, err)
+			}
+		}
 		// Zero out the structure completely upon return. If users of this diskQueue
 		// retain a pointer to it, and we don't remove all references to large
 		// backing slices (various scratch spaces in this struct and children),
@@ -492,12 +499,21 @@ func (d *diskQueue) Close(ctx context.Context) error {
 // It is valid to call rotateFile when the diskQueue is not currently writing to
 // any file (i.e. during initialization). This will simply create the first file
 // to write to.
-func (d *diskQueue) rotateFile(ctx context.Context) error {
+func (d *diskQueue) rotateFile(ctx context.Context) (retErr error) {
 	fName := filepath.Join(d.cfg.GetPather.GetPath(ctx), d.dirName, strconv.Itoa(d.seqNo))
 	f, err := fs.CreateWithSync(d.cfg.FS, fName, bytesPerSync, fs.SQLColumnSpillWriteCategory)
 	if err != nil {
 		return err
 	}
+	defer func() {
+		if retErr != nil {
+			// If we hit an error, then we lose the reference to newly created
+			// file - ensure that it is closed if so.
+			if err = f.Close(); err != nil {
+				retErr = errors.CombineErrors(retErr, err)
+			}
+		}
+	}()
 	d.seqNo++
 
 	if d.serializer == nil {

--- a/pkg/sql/colcontainer/diskqueue_test.go
+++ b/pkg/sql/colcontainer/diskqueue_test.go
@@ -28,6 +28,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/humanizeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/mon"
 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
 	"github.com/stretchr/testify/require"
 )
@@ -178,25 +179,46 @@ func TestDiskQueueCloseOnErr(t *testing.T) {
 	queueCfg, cleanup := colcontainerutils.NewTestingDiskQueueCfg(t, true /* inMem */)
 	defer cleanup()
 
-	serverCfg := &execinfra.ServerConfig{}
-	serverCfg.TestingKnobs.ForceDiskSpill = true
-	diskMon := execinfra.NewLimitedMonitorNoFlowCtx(ctx, testDiskMonitor, serverCfg, nil /* sd */, "test-disk")
-	defer diskMon.Stop(ctx)
-	diskAcc := diskMon.MakeBoundAccount()
-	defer diskAcc.Close(ctx)
+	// Some hard-coded limits assume default batch size.
+	require.NoError(t, coldata.SetBatchSizeForTests(coldata.DefaultColdataBatchSize))
 
-	typs := []*types.T{types.Int}
-	q, err := colcontainer.NewDiskQueue(ctx, typs, queueCfg, &diskAcc, testMemAcc)
-	require.NoError(t, err)
+	rng, _ := randutil.NewTestRand()
+	args := coldatatestutils.RandomVecArgs{Rand: rng}
+	batch := coldatatestutils.RandomBatch(testAllocator, args, types.OneIntCol, coldata.BatchSize(), coldata.BatchSize())
 
-	b := coldata.NewMemBatch(typs, coldata.StandardColumnFactory)
+	// Ensure that each batch is written to a separate file.
+	queueCfg.MaxFileSizeBytes = 1
 
-	err = q.Enqueue(ctx, b)
-	require.Error(t, err, "expected Enqueue to produce an error given a disk limit of one byte")
-	require.Equal(t, pgerror.GetPGCode(err), pgcode.DiskFull, "unexpected pg code")
+	for _, diskLimit := range []int64{
+		2,                             // Disk budget will be exceeded after the first batch.
+		mon.DefaultPoolAllocationSize, // Disk budget will be exceeded after the second batch.
+	} {
+		t.Run(fmt.Sprintf("diskLimit=%s", humanizeutil.IBytes(diskLimit)), func(t *testing.T) {
+			serverCfg := &execinfra.ServerConfig{}
+			serverCfg.TestingKnobs.MemoryLimitBytes = diskLimit
+			diskMon := execinfra.NewLimitedMonitorNoFlowCtx(ctx, testDiskMonitor, serverCfg, nil /* sd */, "test-disk")
+			defer diskMon.Stop(ctx)
+			diskAcc := diskMon.MakeBoundAccount()
+			defer diskAcc.Close(ctx)
 
-	// Now Close the queue, this should be successful.
-	require.NoError(t, q.Close(ctx))
+			typs := []*types.T{types.Int}
+			q, err := colcontainer.NewDiskQueue(ctx, typs, queueCfg, &diskAcc, testMemAcc)
+			require.NoError(t, err)
+
+			err = q.Enqueue(ctx, batch)
+			if diskLimit > 2 {
+				// If we have large disk limit, then enqueuing the first batch
+				// should succeed, but we should get an error on the second one.
+				require.NoError(t, err)
+				err = q.Enqueue(ctx, batch)
+			}
+			require.Error(t, err, "expected Enqueue to produce an error")
+			require.Equal(t, pgerror.GetPGCode(err), pgcode.DiskFull, "unexpected pg code")
+
+			// Now Close the queue, this should be successful.
+			require.NoError(t, q.Close(ctx))
+		})
+	}
 }
 
 // Flags for BenchmarkQueue.

